source/slang/slang-preprocessor.cpp

```diff
@@ -896,280 +1287,763 @@ static Index _parseMacroArgs(
+                // already at its end.
+                //
+                _initCurrentOpStream();
 
-                if (!loc.isValid())
+                // Before we go back to the top of the loop, we need to deal with the
+                // important corner case where `token` might have been an EOF because
+                // the very first op in a macro body had an empty expansion, e.g.:
+                //
+                //      #define TWELVE(X) X 12 X
+                //      TWELVE()
+                //
+                // In this case, the first `X` in the body of the macro will expand
+                // to nothing, so once that op is set up by `_initCurrentOpStrem()`
+                // the `token` we read here will be an EOF.
+                //
+                // The solution is to detect when all preceding ops considered by
+                // this loop have been EOFs, and setting the value to the first
+                // non-EOF token read.
+                //
+                if(token.type == TokenType::EndOfFile)
                 {
-                    // If we don't have a valid source location, don't expand
-                    return;
+                    token = m_currentOpStreams.readToken();
+                    tokenOpIndex = m_macroOpIndex;
                 }
+            }
+            break;
 
-                AdvanceRawToken(preprocessor);
-
-                SourceManager* sourceManager = preprocessor->getSourceManager();
-
-                // Since the location can be overridden by #line directives, use the slower path to get the line number
-                const HumaneSourceLoc humaneSourceLoc = sourceManager->getHumaneLoc(loc);
+        case MacroDefinition::Opcode::TokenPaste:
+            {
+                // The more complicated case is a token paste (`##`).
+                //
+                Index tokenPasteTokenIndex = nextOp.index0;
+                SourceLoc tokenPasteLoc = m_macro->tokens.m_tokens[tokenPasteTokenIndex].loc;
 
-                Token newToken;
+                // A `##` must always appear between two macro ops (whether literal tokens
+                // or macro parameters) and it is supposed to paste together the last
+                // token from the left op with the first token from the right op.
+                //
+                // We will accumulate the pasted token as a string and then re-lex it.
+                //
+                StringBuilder pastedContent;
 
-                StringBuilder buf;
-                if (macro->flavor == PreprocessorMacroFlavor::BuiltinLine)
+                // Note that this is *not* the same as saying that we paste together the
+                // last token the preceded the `##` with the first token that follows it.
+                // In particular, if you have `L ## R` and either `L` or `R` has an empty
+                // expansion, then the `##` should treat that operand as empty.
+                //
```

Comment:
I think this is partly around the confusing question of what is "expansion" or macro invocations, vs. substitution/"expansion" of a macro body. The term "has an empty expansion" here is talking about how a macro "op" expands during the substitution step.

In your first example (`L ## R`) during substitution `L` and `R` are each a raw range of tokens in the macro body. Because the `MacroInvocation` is only responsible for substitution, pasting, stringizing, etc. (but *not* expansion) it does not think of `L` or `R` as something an empty expansion: they are each a sequence of a single raw token.

In your second example (`x ## y`) during substitution `x` and `y` are each an (unexpanded) reference to a macro *parameter*, and as such they will expand to zero or more tokens.

The "empty expansion" case here can really only happen if the preceding "op" was a macro parameter or another token-paste operation involving macro parameters.